src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,303 @@
+ - This instance has the most up to date database
+
+Do you want to proceed? (yes/no): `
+
+const clusterYamlRecoveryComment = `# Member roles can be modified. Unrecoverable nodes should be given the role "spare".
+#
+# "voter" - Voting member of the database. A majority of voters is a quorum.
+# "stand-by" - Non-voting member of the database; can be promoted to voter.
+# "spare" - Not a member of the database.
+#
+# The edit is aborted if:
+# - the number of members changes
+# - the name of any member changes
+# - the ID of any member changes
+# - the address of any member changes
+# - no changes are made
+`
+
+const infoYamlRecoveryComment = `# Verify the ID, address and role of the local node.
+#
+# Cluster members:
+`
+
+const daemonYamlRecoveryComment = `# Verify the name and address of the local node.
+#
+# Cluster members:
+`
+
+var clusterRecoverOpts struct {
+	K8sDqliteStateDir string
+}
+
+
+func newClusterRecoverCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "cluster-recover",
+		Short: "Recover the cluster from this member if quorum is lost",
+		RunE:  clusterRecover,
+	}
+
+	cmd.Flags().StringVar(&clusterRecoverOpts.K8sDqliteStateDir, "state-dir",
+						  "", "k8s-dqlite datastore location")
+
+	return cmd
+}
+
+func removeYamlComments(content []byte) []byte {
+	re := regexp.MustCompile("(?s)#.*?\n")
+	return re.ReplaceAll(content, nil)
+}
+
+func removeYamlCommentsFromFile(path string) error {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	content = removeYamlComments(content)
+
+	return os.WriteFile(path, content, os.FileMode(0o644))
+}
+
+// Add "# " at the beginning of each line
+func yamlCommentLines(content []byte) []byte {
+	re := regexp.MustCompile("(?s)^?\n")
+	return re.ReplaceAll(content, []byte("# "))
+}
+
+func recoverK8sd() error {
+	m, err := microcluster.App(
+		microcluster.Args{
+			StateDir: rootCmdOpts.stateDir,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// The following method parses cluster.yaml and filters out the entries
+	// that are not included in the trust store. Note that in case of k8s-dqlite,
+	// there is no trust store.
+	members, err := m.GetDqliteClusterMembers()
+	if err != nil {
+		return fmt.Errorf("Could not retrieve K8sd cluster members: %w.", err);
+	}
+
+	oldMembersYaml, err := yaml.Marshal(members)
+	if err != nil {
+		return err
+	}
+
+	clusterYamlPath := path.Join(m.FileSystem.DatabaseDir, "cluster.yaml")
+	clusterYamlCommentHeader := fmt.Sprintf("# K8sd cluster configuration\n# (%s)\n", clusterYamlPath)
+	// We aren't passing the file path to our text editor instance since we don't want
+	// the actual content to be written before we get to validate it.
+	clusterYamlContent, err := shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(clusterYamlCommentHeader),
+			[]byte(clusterYamlRecoveryComment),
+			oldMembersYaml))
+	if err != nil {
+		return err
+	}
+
+	infoYamlPath := path.Join(m.FileSystem.DatabaseDir, "info.yaml")
```

Comment:
The cluster.yaml content is commented out, it's there just for informational purposes so that the user won't have to open it in a separate shell.

This is especially useful when copying over the database from a separate node, since you can pick the right id, name and address from cluster.yaml.